Add CiTrace recording support

neobrain opened the pull request:
Adds support for recording CiTraces within Citra (tl;dr: record all GPU commands sent to a game, so that https://github.com/citra/citrace-player can replay them arbitrarily often). The functionality is exposed in a new debugging widget called "CiTrace Recorder".

NOTE: CiTracing needs to be started and halted at GPU breakpoints. The GUI mostly takes care of not allowing the user to start/stop at any other time anyway, but there are still some rough edges.

archshift commented on src/core/hle/service/soc_u.cpp:
```diff
@@ -320,6 +320,9 @@ static void Socket(Service::Interface* self) {
     u32 socket_handle = static_cast<u32>(::socket(domain, type, protocol));
 
+    const int sock_opt = 1;
+    ::setsockopt(socket_handle, SOL_SOCKET, SO_REUSEADDR, &sock_opt, sizeof(sock_opt));
```
Unrelated change?

neobrain replied:
It's required to use the CiTrace player without changing the port after each session, because otherwise creating the listening socket will fail after N runs.

yuriks replied:
Why? And why does SO_RESUEADDR fix this?

neobrain replied:
I guess this should instead be set in the citrace-player, so nevermind (albeit I think it would still be sensible to set this option by default in Citra).

neobrain replied:
The actual is that if an application creates a connection on a particular port and is then forcefully killed and the application is restarted and another attempt for connection is made, there will be an "Address already in use" error in one of the socket setup calls.

yuriks replied:
What @neobrain seems to be hitting is trying to rebind ports when a socket is still in TIME_WAIT (waiting for a clean connection close). SO_REUSEADDR works around this and seems to have no significant disvantages, so it's fine. I think a comment is warranted.

linkmauve commented on src/common/file_util.h:
```diff
@@ -204,6 +204,15 @@ class IOFile : public NonCopyable
+        // TODO: Not supported by gcc 4.8...
+        //static_assert(std::is_trivially_copyable<T>::value, "Given object is not trivially copyable");
+        static_assert(!std::is_pointer<T>::value, "Given object is a pointer");
+        return WriteBytes(&object, sizeof(object)) / sizeof(object);
```
WriteArray would be better suited here.

neobrain replied:
Indeed. I'll change it, and move the static_asserts to that functions instead.

neobrain commented on src/core/hw/gpu.cpp:
```diff
@@ -259,7 +277,16 @@ inline void Write(u32 addr, const T data) {
         if (config.trigger & 1)
         {
             u32* buffer = (u32*)Memory::GetPointer(Memory::PhysicalToVirtualAddress(config.GetPhysicalAddress()));
+            LOG_ERROR(HW_GPU, "Loading list from paddr %x, i.e. vaddr %x",
```
Remove this.

neobrain commented on src/core/hw/gpu.cpp:
```diff
@@ -59,6 +63,8 @@ inline void Write(u32 addr, const T data) {
     addr -= HW::VADDR_GPU;
     u32 index = addr / 4;
 
+    LOG_ERROR(HW_GPU, "Writing %x to %x", (u32)data, addr);
```
Remove this.

bunnei commented on src/core/hw/gpu.cpp:
```diff
@@ -130,12 +138,21 @@ inline void Write(u32 addr, const T data) {
+                // TODO: Figure out something better than this to distinguish between renderer-generated data and user-provided data.
+                if ((config.GetPhysicalInputAddress() + /* input_size */1 <= 0x18000000 || config.GetPhysicalInputAddress() > 0x18600000) &&
+                    (config.GetPhysicalOutputAddress() >= 0x18000000 && config.GetPhysicalOutputAddress() < 0x18600000)) {
+//                    Pica::g_debug_context->recorder->MemoryAccessed(src_pointer, config.input_width * config.input_height * GPU::Regs::BytesPerPixel(config.input_format), config.GetPhysicalInputAddress());
```
?

Subv commented:
What's the status of this PR?

neobrain commented:
Addressed review comments. Reverted the SOC_U change because it's out of place in citra (I'll add it in the actual 3DS application instead).

Please perform your *final* review so that I can rebase and make sure it still works.

bunnei commented:
Please rebase latest master so that CI can build

bunnei commented on src/video_core/command_processor.cpp:
```diff
@@ -182,10 +228,16 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
                 // Send to triangle clipper
                 clipper_primitive_assembler.SubmitVertex(output, Clipper::ProcessTriangle);
             }
+            int i = 0;
```
Unused?

neobrain commented:
Rebased (old version at neobrain@d313426973a155244b6136ff65fd4181b6e3ecde)

bunnei commented on src/video_core/command_processor.cpp:
```diff
@@ -110,12 +110,53 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+                }
+            }
+
+            // map physical start address to size
```
*Map

bunnei commented on src/video_core/command_processor.cpp:
```diff
@@ -110,12 +110,53 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+                    auto it2 = std::next(it);
+                    while (it2 != ranges.end() && it->first + it->second + 32 >= it2->first) {
+                        if (it2->first < it->first)
+                            exit(2);
```
Please don't do this.... assert or something

neobrain replied:
Oh my god, I totally forgot about this.. yeah, definitely.

bunnei commented on src/video_core/clipper.cpp:
```diff
@@ -161,6 +161,7 @@ void ProcessTriangle(OutputVertex &v0, OutputVertex &v1, OutputVertex &v2) {
                   vtx2.screenpos.x.ToFloat32(), vtx2.screenpos.y.ToFloat32(), vtx2.screenpos.z.ToFloat32());
 
         Rasterizer::ProcessTriangle(vtx0, vtx1, vtx2);
+        Rasterizer::ProcessTriangle(vtx0, vtx2, vtx1);
```
What?

bunnei commented on src/core/tracer/tracer.h:
```diff
@@ -0,0 +1,98 @@
+    uint32_t file_offset;
+    uint32_t size;
+    uint32_t physical_address;
+	uint32_t pad;
```
Tab

bunnei commented on src/core/tracer/tracer.h:
```diff
@@ -0,0 +1,98 @@
+        return "CiTr";
+    }
+
+    static uint32_t ExpectedVersion() {
```
Why are you using uint types in this file? u32/u64 please

neobrain replied:
This header is used in citrace_player too, for which I'd rather not use the ugly `uX` types.

bunnei commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,198 @@
+                throw "Failed to write stream element";
+        }
+    } catch(const char* str) {
+        LOG_ERROR(HW_GPU, "Writing CiTrace file failed: %s", str);
```
Should this be a critical?

neobrain replied:
Nope, if this fails, the citrace will not be dumped, but emulation should be able to continue just fine. IMO critical logs should be reserved for errors which are unrecoverable.

yuriks replied:
Agree with neobrain

bunnei commented on src/core/CMakeLists.txt:
```diff
@@ -98,6 +98,10 @@ set(SRCS
+            loader/3dsx.cpp
+            tracer/recorder.cpp
+            core.cpp
+            core_timing.cpp
```
Why are you adding `core.cpp` and `core_timing.cpp` twice?

neobrain replied:
Incorrectly fixed merge conflict, sorry :/

bunnei commented on src/common/file_util.h:
```diff
@@ -204,6 +208,12 @@ class IOFile : public NonCopyable
 
+    template<typename T>
+    size_t WriteObject(const T& object) {
+        static_assert(!std::is_pointer<T>::value, "Given object is a pointer");
```
Should be "Given object is not a pointer" ?

yuriks replied:
Instead of this static assert, wouldn't it be more explicit to simply have the function accept a `T*` in the first place?

Subv replied:
The point of this assert is that you should not pass a pointer to WriteObject

yuriks replied:
Yes, but I perceive it as trying to prevent people from accidentally passing a pointer when they meant to pass the object. I think this could be more directly done by just accepting a pointer in the first place.

neobrain replied:
I can't think of any scenario where anyone would actually want to write a pointer to a file. Hence it's explicitly disallowed.

yuriks replied:
Meh, nevermind.

bunnei commented:
Finished reviewing... still a few mistakes, overall change looks fine though

Subv commented:
I feel like there are already too many lambdas in the code, is there any reason why you're adding these as lambdas instead of normal functions?

neobrain commented on src/video_core/command_processor.cpp:
```diff
@@ -110,12 +110,53 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+
+            // map physical start address to size
+            std::map<u32, u32> accessed_ranges;
+            static auto SimplifyRanges = [](std::map<u32, u32>& ranges) {
```
Consider promoting `accessed_ranges` to a class, of which SimplifyRanges and AddMemoryAccess can be simply made member functions.

neobrain commented:
@Subv as for me personally, it makes no sense to introduce new top-level functions for functionality which is only used in one function anyway. Making it a lambda makes it dead-obvious that the code will only be used in the current function, and nowhere else. It also allows for a more linear code reading, since there's no need to jump between different functions all the time when trying to understand the code.

That said, I think in this particular case accessed_ranges and the two lambdas operating on it would be better suited as a locally implemented utility class.

archshift commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,116 @@
+    // Encode floating point numbers to 24-bit values
+    // TODO: Drop this explicit conversion once we store float24 values bit-correctly internally.
+    std::array<Math::Vec4<uint32_t>, 96> vs_float_uniforms;
+    for (unsigned i = 0; i < 96; ++i)
```
Braces around this? (`for` spans more than one line, and gets a bit messy otherwise)

archshift commented on src/core/tracer/recorder.h:
```diff
@@ -0,0 +1,92 @@
+        std::vector<u8> extra_data;
+
+        /// Optional CRC hash (e.g. for hashing memory regions)
+        boost::crc_32_type::value_type hash;
```
Is there any way you can forward-declare this to avoid adding in the entire boost header in all the files that include this one?

neobrain replied:
I don't think so.

yuriks replied:
You can just use `u32`.

neobrain replied:
@yuriks [citation needed] ?

hrydgard replied:
The type of a CRC32 result must be u32, otherwise it's not a CRC32.

neobrain replied:
@hrydgard While unlikely, future boost versions could decide to change value_type to something like `std::array<uint8_t, 4>`.

I guess I could hardcode u32 in the header, and place a static_assert in the .cpp file to make sure the value_type indeed is u32, but why bother? recorder.h is a rarely included header, hence it's not like compile times are going to explode over including a (comparatively) dependency-lightweight boost header.

yuriks replied:
They could, but they won't, because it wouldn't make any sense.

And it's never the case that doing something "just this once" will explode compilation times, but after you do something "just this once" several times it all adds up.

archshift commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,116 @@
+
+    auto main_widget = new QWidget;
+    auto main_layout = new QVBoxLayout;
+    {
```
Why this extra subscope?

neobrain replied:
We use this kind of style commonly in Qt code, so that each "sub_layout" goes into its own scope. it makes adding multiple sublayouts easier on the eyes, and clearly structures the code.

archshift commented:
All in all, LGTM after you address the outstanding comments.

Subv commented on src/video_core/command_processor.cpp:
```diff
@@ -110,12 +110,53 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
                     // TODO: Implement some sort of vertex cache!
+                    if (g_debug_context && Pica::g_debug_context->recorder) {
+                        int size = index_u16 ? 2 : 1;
+                        AddMemoryAccess(accessed_ranges, base_address + index_info.offset + size*index, size);
```
How much does this call affect performance?

neobrain replied:
Not at all, unless citracing is active (in which case we don't care, I would say).

Subv commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,116 @@
+        for (unsigned comp = 0; comp < 3; ++comp)
+            vs_float_uniforms[i][comp] = nihstro::to_float24(Pica::VertexShader::GetFloatUniform(i)[comp].ToFloat32());
+
+    auto recorder = new CiTrace::Recorder((u32*)&GPU::g_regs, 0x700, nullptr, 0, (u32*)&Pica::registers, 0x300,
```
It doesn't seem like you're saving the default attributes to the CiTrace

neobrain replied:
Good point!

neobrain commented:
All comments should've been addressed by now.

Some new code was added to 1) record default vertex attributes 2) properly enable/disable the citrace widget upon stopping/starting emulation.

neobrain commented:
There's some potential for merge conflicts with regards to the pull requests  #776 and  #789. I suggest merging #776 first, then this PR, then #789, and deal with any merge conflicts in that same order.

neobrain commented:
Rebased.

(old version was f8d8e5ff620)

neobrain commented:
Chances are the hardware renderer needs changes for this to work. CC @tfarley.

neobrain commented:
Chances are the hardware renderer needs changes for this to work. CC @tfarley.

In any case, I don't consider this blocking this PR.

Subv commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,148 @@
+    // Encode floating point numbers to 24-bit values
+    // TODO: Drop this explicit conversion once we store float24 values bit-correctly internally.
+    std::array<Math::Vec4<uint32_t>, 16> default_attributes;
+    for (unsigned i = 0; i < 16; ++i) {
```
Why do these loops have brackets and the vs_float_uniforms loops don't?

Subv commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,148 @@
+        for (unsigned comp = 0; comp < 3; ++comp)
+            vs_float_uniforms[i][comp] = nihstro::to_float24(Pica::g_state.vs.uniforms.f[i][comp].ToFloat32());
+
+    auto recorder = new CiTrace::Recorder((u32*)&GPU::g_regs, 0x700, nullptr, 0, (u32*)&Pica::g_state.regs, 0x300,
```
What do 0x700 and 0x300 mean?

neobrain replied:
static_cast is not a legal method to cast between unrelated pointer types. I guess I should use reinterpret_cast, though, even though that's not valid either, but at least compiles with strict aliasing rules disabled.

Subv replied:
Why are you not passing the LCD registers?

neobrain replied:
Heh, I was about to push a fix for that ;)

At the time I wrote this code, LCD hadn't been implemented.

Subv commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    // Allow starting tracing before emulation is started
```
What does this comment mean? We're in OnEmulationStopping

neobrain replied:
I'll clarify it.

Subv commented on src/common/file_util.h:
```diff
@@ -182,6 +182,10 @@ class IOFile : public NonCopyable
     {
+        static_assert(std::is_standard_layout<T>::value, "Given array does not consist of standard layout objects");
+        // TODO: gcc 4.8 does not support is_trivially_copyable, but we really should check for it here.
+        //static_assert(std::is_trivially_copyable<T>::value, "Given array does not consist of trivially copyable objects");
```
Dolphin has a workaround for this, see https://github.com/dolphin-emu/dolphin/pull/2440/files

neobrain replied:
I guess we could use that, but at the same time I think it's not particularly critical to address it in this PR, is it?

Subv commented on src/video_core/command_processor.cpp:
```diff
@@ -111,12 +111,55 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+                }
+            }
+
+            class {
```
What?

neobrain replied:
The case in the switch of this function is literally the only place where this class is used. Moving it outside this function would make it appear like a more generic class and hence isn't any less confusing than having a local class.

yuriks replied:
Agreed with @Subv. This breaks up the reading flow for something that is only used by tracing, which is an incidental feature that you'll almost never be paying attention to when reading this class.

Subv commented on src/video_core/command_processor.cpp:
```diff
@@ -111,12 +111,55 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
                     // TODO: Implement some sort of vertex cache!
+                    if (g_debug_context && Pica::g_debug_context->recorder) {
+                        int size = index_u16 ? 2 : 1;
+                        memory_accesses.AddAccess(base_address + index_info.offset + size * index, size);
```
How much does the SimplifyRanges operation impact the running time of this loop?

neobrain replied:
Not at all unless citracing is active.

yuriks replied:
Isn't there a way to know the size of the vertex buffer beforehand, and then just uploading the whole thing, instead of bothering with all this?

yuriks replied:
(Even if there isn't, the simplification logic can be greatly simplified by just keep tracking of a single min,max address range for the whole draw operation. This will be efficient since vertices already have locality anyway.)

yuriks replied:
(Even if there isn't, the logic can be greatly simplified by just keep tracking of a single min,max address range for the each loader and then merging them if they overlap. This handles the any combination of a single interleaved or many partially interleaved or dense buffers. This will be efficient since even indexed vertices already have locality inside their own buffers anyway.)

neobrain commented:
Addressed another round of comments.

* It was suggested to use a C++-style cast in a situation where I'm casting a Regs* pointer to a u32* pointer. This is not actually possible without violating strict aliasing rules, and while a `reinterpret_cast` compiles and behaves perfectly fine for that purpose, it seems wrong to me to use a cast implying a particular intention when using that cast is not actually valid.
* It was suggested to move a local helper class into the global scope of the translation unit. I still think it makes more sense to have the class be a locally defined one because its usage is highly localized (to a case within a switch statement in a function).

yuriks commented:
Build is failing on all platforms...

neobrain commented:
.. fixed :/

yuriks commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,153 @@
+
+GraphicsTracingWidget::GraphicsTracingWidget(std::shared_ptr<Pica::DebugContext> debug_context,
+                                             QWidget* parent)
+    : BreakPointObserverDock(debug_context, tr("CiTrace Recorder"), parent) {
```
Missing "™" after "CiTrace".

yuriks commented on src/citra_qt/debugger/graphics_tracing.h:
```diff
@@ -0,0 +1,32 @@
+signals:
+    void SetStartTracingButtonEnabled(bool enable);
+    void SetStopTracingButtonEnabled(bool enable);
+    void SetAbortTracingButtonEnabled(bool enable);
```
Wtf is the point of these signals? Why aren't they just a function???

neobrain replied:
Making them functions requires keeping around pointers to the respective buttons, which is bad style IMO.

neobrain replied:
Also, it actually requires implementing three further functions instead of just calling connect once for each of them.

yuriks replied:
Uh, no it does not require implementing three functions. It requires you to move three pointers variables from being function local to being class members and that's it. And it's not bad style, it's literally what the Qt UI designer does.

yuriks commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,153 @@
+}
+
+void GraphicsTracingWidget::StartRecording() {
+    auto context = context_weak.lock();
```
`context` isn't checked for being null before being used, in case acquiring the `weak_ptr` fails. The same error is made in all other uses in the PR.

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+                   u32* vs_float_uniforms,  u32 vs_float_uniforms_size,
+                   u32* gs_program_binary,  u32 gs_program_binary_size,
+                   u32* gs_swizzle_data,    u32 gs_swizzle_data_size,
+                   u32* gs_float_uniforms,  u32 gs_float_uniforms_size)
```
This is simply terrible. A function with a zillion parameters is very error prone to use and confusing to read, especially seeing as they all have the same type, there's no way to know if you got the order right or to modify the parameter list without breaking users unknowingly. Define a parameters struct that you fill in and pass to the function instead.

yuriks replied:
Also, considering that the treatment of all this data seems to be uniform, consider if you can't simply use an array or something.

yuriks commented on src/citra_qt/debugger/graphics_tracing.cpp:
```diff
@@ -0,0 +1,153 @@
+    auto context = context_weak.lock();
+
+    if (filename.isEmpty()) {
+        // If no filename was given, keep recording
```
"If the user cancelled the dialog, keep recording."

yuriks commented on src/core/hw/gpu.cpp:
```diff
@@ -75,25 +79,27 @@ inline void Write(u32 addr, const T data) {
-                    ptr[1] = config.value_24bit_g;
-                    ptr[2] = config.value_24bit_b;
+        if (config.trigger) {
+            if (config.address_start) { // Some games pass invalid values here
```
Why split this?

neobrain replied:
Because some program code should be executed even if address_start is zero, cf below.

yuriks replied:
Which code? None of the code that was already there should be executed if the transfer isn't configured, and you didn't add any new code in this scope.

neobrain replied:
Just read the code? I'm not sure where the issue is, really. if config.trigger and config.address_start are both true, all code is being enabled, and if only config.trigger is true, only part of it is enabled. That's what happens on hardware. No idea why they use such odd configurations as trigger=true and address_start=0.

yuriks replied:
If no copy happened we should neither log (just turns into a misleading red herring) or signal a hw renderer flush. I don't think an interrupt is fired in that case either, from my earlier experiments (and @Subv also getting similar results.). Resetting the config register might be valid though.

neobrain replied:
Fair enough. Resetting the config register is the only thing I actually verified explicitly.

neobrain commented:
There is an annoying trend in our PR review process, which I would like to clarify here. Case in point, the review made by @yuriks is IMO beyond any reasonable timeline. It's been more than a month since I opened this PR. No offense, but I don't feel like going through major iterations over something which was already public for weeks.

Apparently there was a misunderstanding as to how I'm addressing review comments - people seem to think that if one person gave his review on a PR, those review comments should be addressed before the next person can jump in to give his review. This is a really inefficent process though, since even in the better cases it requires multiple iterations on the PR (instead of just having the PR author address all comments in one go).

Let me make this clear for once: When opening PRs, my expectation is that everyone who wants to have a say on things will do so immediately without waiting for other events (delaying review is what we have the RFC tagfor). I'm not willing to put up with an inefficient review process which by design requires me to go through multiple iterations beyond the initial PRing phase. Sorry if this comes across a bit demand-ish, but it simply will not work any other way.

I will look over the "late review comments" which have been added earlier today and consider fixing them, but in the future my tendency will go towards ignoring such late comments unless their background has a critical impact on software quality.

yuriks commented on src/core/hw/hw.cpp:
```diff
@@ -13,7 +13,7 @@ namespace HW {
 template <typename T>
 inline void Read(T &var, const u32 addr) {
-    switch (addr & 0xFFFFF000) {
+    switch (addr & 0xFFFF0000) {
```
I assume you changed the mask because the GPU register range spans more than a page (0x1000). However all other registers (see http://3dbrew.org/wiki/IO) are on individual pages, so masking like this will cause writes to other ranges to fail. You should instead just add multiple cases for the GPU.

yuriks commented on src/core/tracer/citrace.h:
```diff
@@ -0,0 +1,101 @@
+
+namespace CiTrace {
+
+// NOTE: Things are stored in little-endian
```
Then shouldn't you be using `u32_le` types?

yuriks commented on src/core/tracer/citrace.h:
```diff
@@ -0,0 +1,101 @@
+
+// NOTE: Things are stored in little-endian
+
+#pragma pack(1)
```
This doesn't seem necessary at all, since the entire struct has natural padding already.

neobrain replied:
Are any of these "natural alignment"s guaranteed by spec? If not, I'd rather keep the `#pragma pack(1)` to indeed guarantee it.

yuriks replied:
They're not, but neither is `#pragma pack` or that unaligned accesses should work, so it comes down to the platforms you're supporting.

neobrain replied:
Why should the two things you mentioned (the latter which I'm not sure what it's referring to exactly) not work?

hrydgard replied:
If things are naturally aligned, packing is indeed guaranteed and no pragma is necessary.

neobrain replied:
FYI, I tried removing this and it completely broke citrace recordings. I guess this is just due to some member of the code breaking natural alignment and I'm open to suggestions as to where that member is, but for now I don't really feel like finding the culprit and will just keep the alignment specifier.

yuriks commented on src/core/tracer/citrace.h:
```diff
@@ -0,0 +1,101 @@
+};
+
+struct CTStreamElement {
+    CTStreamElementType type;
```
Do we also want to store timing info, in case real-time trace replays are desirable?

neobrain replied:
I don't think there's a reason to store this sort of information. In any case, it can be added in a later version if it's deemed useful after all.

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+}
+
+void Recorder::Finish(const std::string& filename) {
+    // TODO: Make sure this is only called when the gpu thread is paused!
```
Have you made sure? Alternatively, should we try to assert that somehow?

neobrain replied:
Indeed, I have made sure it's only called when the GPU thread is paused. This TODO should be removed.

I don't think there's a good way to assert for this condition in general, though.

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+
+        // Write initial state
+        written = file.WriteArray(gpu_registers.data(), gpu_registers.size());
+        if (written != gpu_registers.size() || file.Tell() != initial.lcd_registers)
```
Doesn't `IOFile` already keep track of a "failed" flag that you can simply check at the end to know if something went wrong when writing? Additionally, asserting the cursor position after every read is redundant with the code above which calculates the offsets.

yuriks replied:
(I you want to ensure you're at the right place I'd rather you seek beforehand so that at least the calculation of the offsets above can be changed without affecting this whole chain of tell checks.)

yuriks commented on src/core/tracer/citrace.h:
```diff
@@ -0,0 +1,101 @@
+    uint32_t file_offset;
+    uint32_t size;
+    uint32_t physical_address;
+    uint32_t pad;
```
Why the manual pad?

neobrain replied:
To make sure the sizes of the CTMemoryLoad, CTRegisterWrite, and CTStreamElement structs are equal.

yuriks replied:
But why is that important? The union below will already take the max size of all it's members.

neobrain replied:
I don't remember because it's been ages since I wrote this code, but I did have a good reason for it.

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+    stream.push_back(element);
+}
+
+template void Recorder::RegisterWritten(u32,u8);
```
Space after the comma

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+    if (!element.uses_existing_data) {
+        element.extra_data.resize(size);
+        memcpy(element.extra_data.data(), data, size);
+        memory_regions.insert({result.checksum(), 0}); // file offset will be initialized in Finish()
```
Why not simply use `element.hash` instead of `result.checksum()`? (Less redundancy, potentially faster if `checksum()` does a non-trivial amount of work.)

yuriks replied:
Is there any point in inserting the dummy zero here? Nothing seem to actually depends on this. (Be aware that operator[] automatically inserts a blank element in the map if the key isn't found.)

yuriks replied:
Also, instead of `.insert({foo, bar})` you can just use `.emplace(foo, bar)`.

neobrain replied:
The point of inserting the dummy zero here is to make sure the checksum can be found via `.find()` in successive calls to AccessedMemory.

yuriks replied:
Ah, ok.

yuriks commented on src/core/tracer/recorder.cpp:
```diff
@@ -0,0 +1,206 @@
+    result.process_bytes(data, size);
+    element.hash = result.checksum();
+
+    element.uses_existing_data = (memory_regions.find(result.checksum()) != memory_regions.end());
```
Same comment as below.

yuriks commented on src/core/tracer/recorder.h:
```diff
@@ -0,0 +1,95 @@
+     * Internal cache which maps hashes of memory contents to file offsets at which those memory
+     * contents are stored.
+     */
+    std::map<boost::crc_32_type::value_type /*hash*/, u32 /*file_offset*/> memory_regions;
```
Unless you need the data in sorted order, use `std::unordered_map`.

yuriks commented on src/core/tracer/recorder.h:
```diff
@@ -0,0 +1,95 @@
+     * @note Use this whenever the GPU is about to access a particular memory region.
+     * @note The implementation will make sure to minimize redundant memory updates.
+     */
+    void MemoryAccessed(const u8* data, u32 size, u32 physical_address);
```
The name of this func just sounds really weird and vague to me, especially when the documentation explicitly says what it actually does.

neobrain replied:
Could you suggest a better name?

yuriks replied:
Something like `CaptureMemoryRegion`, `RecordRegisterWrite` and `RecordFrameFinished`.

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -136,10 +179,17 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+                        u32 source_addr = vertex_attribute_sources[i] + vertex_attribute_strides[i] * vertex + comp * vertex_attribute_element_size[i];
+                        const u8* srcdata = Memory::GetPhysicalPointer(source_addr);
+
+                        if (g_debug_context && Pica::g_debug_context->recorder) {
```
`g_debug_context` and then `Pica::g_debug_context`. Make it consistent.

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -111,12 +111,55 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
 
                 if (is_indexed) {
                     // TODO: Implement some sort of vertex cache!
+                    if (g_debug_context && Pica::g_debug_context->recorder) {
```
`Pica::` only used on the second access.

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -136,10 +179,17 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+                        if (g_debug_context && Pica::g_debug_context->recorder) {
+                            memory_accesses.AddAccess(source_addr,
+                                    (vertex_attribute_formats[i] == Regs::VertexAttributeFormat::FLOAT) ? 4
+                                    : (vertex_attribute_formats[i] == Regs::VertexAttributeFormat::SHORT) ? 2 : 1);
```
Just add the whole vertex instead of using fine-grained detection like this?

yuriks commented:
Pass finished.

neobrain commented:
Updated again. I'll merge this branch as soon as I have tested it once again to make sure it still works.

yuriks commented on src/core/tracer/citrace.h:
```diff
@@ -0,0 +1,101 @@
+struct CTStreamElement {
+    CTStreamElementType type;
+
+    union {
```
Missing padding is before this union. The union has 64-bit alignment due to CTRegisterWrite, so you need to add a u32 of padding here.
